File: grammar.py
import typing

# Man, I forgot a lot of my compiler design theory since college...
# https://www.tutorialspoint.com/compiler_design/compiler_design_syntax_analysis.htm
# https://stackoverflow.com/questions/20317198/purpose-of-first-and-follow-sets-in-ll1-parsers

grammar = """
  Root                = FuncList
  FuncList            = Func FuncList'
  FuncList'           = FuncList | Nil
  Func                = Type $Id "(" Func'
  Func'               = ArgList ")" StatementList | ")" StatementList
  StatementList       = Statement ";" | "{" Statement ";" StatementList "}"
  Statement           = "return" Expression | $Id Statement'
  
  //                    Decl  Assign 
  Statement'          = $Id | Expression'' "=" Expression 
  Type                = $Id
  ArgList             = Arg ArgList'
  ArgList'            = "," ArgList | Nil
  Arg                 = Type $Id
  Expression          = $Id Expression' | $Int32 Expression'
  Expression'         = Expression'' | Nil
  Expression''        = "==" Expression
"""


def str_to_basic(str_table):
    rules_temp = {}

    for line in str_table.splitlines():
        line = line.split("//")[0]
        line = line.split()

        if len(line) == 0:
            continue

        productions = []
        acc = []

        assert line[1] == "="
        current = line[2:]
        while len(current):
            if current[0] == '|':
                productions.append(acc)
                acc = []
            else:
                acc.append(current[0])
            current = current[1:]

        if len(acc):
            productions.append(acc)

        rules_temp[line[0]] = productions

    return rules_temp


def make_grammar(str_table: str) -> typing.Dict[str, "NonTerminal"]:
    rules_temp = str_to_basic(str_table)
    rules: typing.Dict[str, NonTerminal] = {key: NonTerminal(key) for key in rules_temp.keys()}

    for name in rules:
        for prod_index in range(len(rules_temp[name])):
            production = rules_temp[name][prod_index]
            if len(production) > 1 and 'Nil' in production:
                raise Exception('Nil can only appear alone')
            if production == ['Nil'] and prod_index != len(rules_temp[name]) -1:
                raise Exception('Nil has to be the last production')

            for i in range(len(production)):
                if production[i] != 'Nil' and production[i][0].isupper():
                    production[i] = rules[production[i]]

            rules[name].productions.append(production)

    return rules


Production = typing.List[typing.Union[str, "NonTerminal"]]


class NonTerminal:
    def __init__(self, name: str):
        self.name = name
        self.productions: typing.List[Production] = []

    def rules_str(self):
        strs = []
        for production in self.productions:
            strs += [str(x) for x in production]
        return f"{self} = {' | '.join(strs)}"

    def __str__(self):
        return f"NT({self.name})"

    def __repr__(self):
        return self.__str__()


class Grammar:
    def __init__(self, str_table: str):
        self.rules: typing.Dict[str, NonTerminal] = make_grammar(str_table)

    def can_be_nil(self, name: str) -> bool:
        for production in self.rules[name].productions:
            if production[0] == 'Nil':
                return True
            elif isinstance(production[0], NonTerminal):
                test_prod = production
                while True:
                    if not isinstance(test_prod[0], NonTerminal):
                        break
                    if not self.can_be_nil(test_prod[0].name):
                        break

                    test_prod = test_prod[1:]

                    if len(test_prod) == 0:
                        return True

        return False

    def _first_internal(self, name: str) -> typing.List[typing.List[str]]:
        retval = []
        for production in self.rules[name].productions:
            prod_temp = production
            prod_firsts = []
            while len(prod_temp):
                if not isinstance(prod_temp[0], NonTerminal):
                    if prod_temp[0] != 'Nil':
                        prod_firsts.append(prod_temp[0])
                    break

                for x in self.first(prod_temp[0].name):
                    for item in x:
                        prod_firsts.append(item)

                if not self.can_be_nil(prod_temp[0].name):
                    break

                prod_temp = prod_temp[1:]

            if len(prod_firsts):
                retval.append(prod_firsts)

        return retval

    def first(self, name: str) -> typing.List[typing.List[str]]:
        retval = self._first_internal(name)
        tmp = set({})
        for x in retval:
            for item in x:
                if item in tmp:
                    raise Exception("duplicate first")
                tmp.add(item)
        return retval

    def _pad(self, name: str) -> str:
        max_len = 0
        for other_name in self.rules:
            max_len = max(len(other_name), max_len)
        return ' ' * (max_len - len(name))

    def str_first(self, name: str) -> str:
        strs = [f"{name} {self._pad(name)}="]
        for prod_firsts in self.first(name):
            for item in prod_firsts:
                strs.append(item)
            strs.append('|')
        strs = strs[:-1]
        return " ".join(strs)

    def follow(self, name: str, _visited=None) -> typing.Set[str]:
        if _visited is None:
            _visited = set()

        if name in _visited:
            return set()

        _visited.add(name)

        follows: typing.Set[str] = set()
        nt = self.rules[name]

        if name == 'Root':
            follows.add('$End')

        for other_name in self.rules:
            other_nt = self.rules[other_name]
            for production in other_nt.productions:
                for i in range(len(production)):
                    if nt is production[i]:
                        prod_temp = production[i:]
                        while True:
                            if len(prod_temp) == 1:
                                if isinstance(prod_temp[0], str) or prod_temp[0] is nt or self.can_be_nil(prod_temp[0].name):
                                    follows = follows.union(self.follow(other_nt.name, _visited))
                                break
                            else:
                                next_symbol = prod_temp[1]
                                if isinstance(next_symbol, NonTerminal):
                                    for x in self.first(next_symbol.name):
                                        for item in x:
                                            follows.add(item)
                                    if not self.can_be_nil(next_symbol.name):
                                        break
                                else:
                                    follows.add(next_symbol)
                                    break

                            prod_temp = prod_temp[1:]
        return follows

    def str_follow(self, name: str) -> str:
        strs = [f"{name} {self._pad(name)}="]
        for item in self.follow(name):
            strs.append(item)
        return " ".join(strs)


def test_can_be_nil():
    rules = Grammar("""
        Root = "3" Y
        Y = A | "1"
        A = "2" | Nil
        """)
    assert not rules.can_be_nil('Root')
    assert rules.can_be_nil('A')
    assert rules.can_be_nil('Y')

    rules = Grammar("""
        Root = "x" Y
        Y = A B C D
        A = "1" | Nil
        B = "2" | Nil
        C = "3" | Nil
        D = "4" | Nil
        """)
    assert not rules.can_be_nil('Root')
    assert rules.can_be_nil('A')
    assert rules.can_be_nil('B')
    assert rules.can_be_nil('C')
    assert rules.can_be_nil('D')
    assert rules.can_be_nil('Y')

    rules = Grammar("""
        Root = "x" Y
        Y = A B C D
        A = "1" | Nil
        B = "2" | Nil
        C = "3"
        D = "4" | Nil
        """)
    assert not rules.can_be_nil('Root')
    assert rules.can_be_nil('A')
    assert rules.can_be_nil('B')
    assert not rules.can_be_nil('C')
    assert rules.can_be_nil('D')
    assert not rules.can_be_nil('Y')


def test_first():
    rules = Grammar("""
        Root = "3" Y
        Y = A | "1"
        A = "2" | Nil
        """)
    assert rules.first('Root') == [['"3"']]
    assert rules.first('A') == [['"2"']]
    assert rules.first('Y') == [['"2"'], ['"1"']]

    rules = Grammar("""
        Root = "3" Y
        Y = A | "1"
        A = "2" | "4"
        """)
    assert rules.first('Root') == [['"3"']]
    assert rules.first('A') == [['"2"'], ['"4"']]
    assert rules.first('Y') == [['"2"', '"4"'], ['"1"']]

    rules = Grammar("""
        Root = "x" Y
        Y = A B C D
        A = "1" | Nil
        B = "2" | Nil
        C = "3"
        D = "4" | Nil
        """)
    assert rules.first('A') == [['"1"']]
    assert rules.first('B') == [['"2"']]
    assert rules.first('C') == [['"3"']]
    assert rules.first('D') == [['"4"']]
    assert rules.first('Y') == [['"1"', '"2"', '"3"']]

    rules = Grammar("""
        Root = "x" Y
        Y = A B C D
        A = "1" | Nil
        B = "2" | Nil
        C = "3" | Nil
        D = "4" | Nil
        """)
    assert rules.first('A') == [['"1"']]
    assert rules.first('B') == [['"2"']]
    assert rules.first('C') == [['"3"']]
    assert rules.first('D') == [['"4"']]
    assert rules.first('Y') == [['"1"', '"2"', '"3"', '"4"']]


def test_follow():
    rules = Grammar("""
        Root = A Y
        Y = A | "1"
        A = "2" | "3"
        """)

    assert rules.follow('Root') == {'$End'}
    assert rules.follow('A') == {'"1"', '"2"', '"3"', '$End'}
    assert rules.follow('Y') == {'$End'}

    rules = Grammar("""
        Root = "x" Y
        Y = A B C D
        A = "1" | Nil
        B = "2" | Nil
        C = "3"
        D = "4" | Nil
        """)

    assert rules.follow('Root') == {'$End'}
    assert rules.follow('Y') == {'$End'}
    assert rules.follow('A') == {'"2"', '"3"'}
    assert rules.follow('B') == {'"3"'}
    assert rules.follow('C') == {'"4"', '$End'}


def test():
    test_can_be_nil()
    test_first()
    test_follow()
    print("tests passed!")


def main():
    rules = Grammar(grammar)

    print("Firsts:")
    for name in rules.rules:
        print('   ', rules.str_first(name))

    print()
    print("Follows:")
    for name in rules.rules:
        print('   ', rules.str_follow(name))


if __name__ == "__main__":
    test()
    main()

File: GrammarTool/Grammar.cpp
#include <functional>
#include <sstream>
#include "Grammar.hpp"
#include "../StringUtil.hpp"


// Man, I forgot a lot of my compiler design theory since college...
// https://www.tutorialspoint.com/compiler_design/compiler_design_syntax_analysis.htm
// https://stackoverflow.com/questions/20317198/purpose-of-first-and-follow-sets-in-ll1-parsers

static std::vector<std::string> tokenise(std::string_view input)
{
  std::vector<std::string> tokens;

  bool inQuotedString = false;
  int32_t angleBracketDepth = 0;
  bool inComment = false;
  bool inCodeInsert = false;
  std::string accumulator;

  auto breakToken = [&]()
  {
    if (accumulator.empty())
      return;

    tokens.emplace_back(std::move(accumulator));
    accumulator.clear();
  };

  auto advance = [&](size_t chars)
  {
    debug_assert(input.size() >= chars);
    input = std::string_view(input.data() + chars, input.size() - chars);
  };

  while (!input.empty())
  {
    if (inComment)
    {
      if (input[0] == '\n')
        inComment = false;
      advance(1);
      continue;
    }

    if (!inQuotedString && !inComment && angleBracketDepth == 0 && !inCodeInsert)
    {
      if (Str::isSpace(input[0]))
      {
        breakToken();
        advance(1);
        continue;
      }

      if (input[0] == '|' || input[0] == '=' || input[0] == ';')
      {
        breakToken();
        accumulator.push_back(input[0]);
        breakToken();
        advance(1);
        continue;
      }

      if (input.starts_with("//"))
      {
        advance(2);
        inComment = true;
        continue;
      }

      if (input[0] == '<')
      {
        breakToken();
        advance(1);
        angleBracketDepth = 1;
        accumulator.push_back('<');
        continue;
      }

      if (input.starts_with("{{"))
      {
        breakToken();
        advance(2);
        inCodeInsert = true;
        accumulator += "{{";
        continue;
      }
    }

    if (accumulator.empty() && input[0] == '"')
    {
      advance(1);
      inQuotedString = true;
      accumulator.push_back('"');
      continue;
    }

    if (inQuotedString && input[0] == '"')
    {
      advance(1);
      inQuotedString = false;
      accumulator.push_back('"');
      breakToken();
      continue;
    }

    if (angleBracketDepth)
    {
      if (input[0] == '>')
      {
        angleBracketDepth--;

        if (angleBracketDepth == 0)
        {
          advance(1);
          accumulator.push_back('>');
          breakToken();
          continue;
        }
      }
      else if (input[0] == '<')
      {
        angleBracketDepth++;
      }
    }

    if (inCodeInsert && input.starts_with("}}"))
    {
      advance(2);
      inCodeInsert = false;
      accumulator += "}}";
      breakToken();
      continue;
    }

    accumulator.push_back(input[0]);
    advance(1);
  }

  return tokens;
}


struct GrammarResult
{
  std::unordered_map<std::string, NonTerminal> rules;
  std::vector<std::string> keys;
};

static GrammarResult make_grammar(const std::string& str_table)
{
  GrammarResult rules;
  {
    std::vector<std::string> tokens = tokenise(str_table);

    int32_t ruleStartIndex = 0;
    std::string currentRuleName;
    Production accumulator;

    std::string codeInsert;

    auto breakAccumulator = [&]()
    {
      if (!accumulator.empty())
      {
        if (!codeInsert.empty())
        {
          accumulator[accumulator.size()-1].codeInsertAfter = std::move(codeInsert);
          codeInsert.clear();
        }
        rules.rules[currentRuleName].productions.emplace_back(std::move(accumulator));
      }
      accumulator.clear();
    };

    for (int32_t i = 0; i < int32_t(tokens.size()); i++)
    {
      if (i == ruleStartIndex)
      {
        release_assert(tokens[i] != ";");
        release_assert((i < int32_t(tokens.size()) - 1 && tokens[i+1] == "=") ||
                       (i < int32_t(tokens.size()) - 2 && tokens[i+1][0] == '<' && tokens[i+2] == "="));
        rules.rules[tokens[i]] = NonTerminal { .name = tokens[i] };
        rules.keys.emplace_back(tokens[i]);
        currentRuleName = tokens[i];
        ruleStartIndex = i;
        continue;
      }

      if (tokens[i] == "=")
      {
        release_assert((i == ruleStartIndex + 1) ||
                       (i == ruleStartIndex + 2 && tokens[i-1][0] == '<'));
        continue;
      }

      if (tokens[i] == ";")
      {
        breakAccumulator();
        ruleStartIndex = i + 1;
        continue;
      }

      if (tokens[i] == "|")
      {
        breakAccumulator();
        continue;
      }

      if (tokens[i].starts_with("<"))
      {
        release_assert(tokens[i][tokens[i].length()-1] == '>');
        release_assert(i == ruleStartIndex + 1);
        rules.rules[currentRuleName].returnType = tokens[i].substr(1, tokens[i].length()-2);
        continue;
      }

      if (tokens[i].starts_with("{{"))
      {
        release_assert(tokens[i].ends_with("}}"));
        codeInsert = tokens[i].substr(2, tokens[i].length()-4);
        continue;
      }

      ProductionItem& item = accumulator.emplace_back(tokens[i]);
      if (!codeInsert.empty())
      {
        item.codeInsertBefore = std::move(codeInsert);
        codeInsert.clear();
      }
    }
  }

  for (const std::string& name: rules.keys)
  {
    for (int32_t prod_index = 0; prod_index < int32_t(rules.rules[name].productions.size()); prod_index++)
    {
      Production& production = rules.rules[name].productions[prod_index];

      if (production.size() > 1)
      {
        for (const auto& item: production)
        {
          if (item == "Nil")
            release_assert(false && "Nil can only appear alone");
        }
      }

      if (production.size() == 1 && production[0] == "Nil" && prod_index != int32_t(rules.rules[name].productions.size()) - 1)
        release_assert(false && "Nil has to be the last production");

      for (int32_t i = 0; i < int32_t(production.size()); i++)
      {
        if (production[i] != "Nil" && isupper(production[i].str()[0]))
          production[i].setNonTerminal(rules.rules[production[i].str()]);
      }
    }
  }

  return rules;
}


Grammar::Grammar(const std::string& str_table)
{
  GrammarResult result = make_grammar(str_table);
  this->rules = std::move(result.rules);
  this->keys = std::move(result.keys);
}

bool Grammar::can_be_nil(const std::string& name) const
{
  for (const Production& production : this->rules.at(name).productions)
  {
    if (production[0] == "Nil")
      return true;

    if (production[0].isNonTerminal())
    {
      Production test_prod = production;
      while (true)
      {
          if (!test_prod[0].isNonTerminal())
              break;
          if (!this->can_be_nil(test_prod[0].nonTerminal().name))
              break;

          test_prod.erase(test_prod.begin());

          if (test_prod.empty())
              return true;
      }
    }
  }

  return false;
}

std::vector<std::vector<std::string>> Grammar::first(const std::string& name) const
{
  std::vector<std::vector<std::string>> retval;
  for (const Production& production : this->rules.at(name).productions)
  {
    std::vector<std::string> prod_firsts;
    auto prod_firsts_append = [&](const std::string& sym)
    {
      if (sym == "Nil" && std::find(prod_firsts.begin(), prod_firsts.end(), "Nil") != prod_firsts.end())
        return;
      prod_firsts.emplace_back(sym);
    };

    for (int32_t i = 0; i < int32_t(production.size()); i++)
    {
      if (!production[i].isNonTerminal())
      {
        prod_firsts_append(production[i].str());
        break;
      }

      const NonTerminal& nonTerminal = production[i].nonTerminal();

      for (const std::vector<std::string>& x : this->first(nonTerminal.name))
      {
        for (const std::string& item: x)
          prod_firsts_append(item);
      }

      if (!this->can_be_nil(nonTerminal.name))
        break;
    }

    if (!prod_firsts.empty())
      retval.emplace_back(std::move(prod_firsts));
  }

  // check for duplicates
  {
    std::unordered_set<std::string> tmp;
    for (const std::vector<std::string>& x: retval)
    {
      for (const std::string& item: x)
      {
        release_assert(!tmp.contains(item) && "duplicate first!");
        tmp.insert(item);
      }
    }
  }

  return retval;
}

std::unordered_set<std::string> Grammar::follow(const std::string& _name) const
{
  std::unordered_set<std::string> _visited;
  std::function<std::unordered_set<std::string>(const std::string&)> followInner = [&](const std::string& name) -> std::unordered_set<std::string>
  {
    if (_visited.contains(name))
      return {};

    _visited.insert(name);

    std::unordered_set<std::string> follows;
    const NonTerminal& nt = this->rules.at(name);

    for (const auto& pair : this->rules)
    {
      const NonTerminal& other_nt = pair.second;
      for (const Production& production: other_nt.productions)
      {
        for (int32_t i = 0; i < int32_t(production.size()); i++)
        {
          if (production[i] == nt)
          {
            Production prod_temp(production.begin() + i, production.end());
            while (true)
            {
              if (prod_temp.size() == 1)
              {
                if (prod_temp[0].isStr() || prod_temp[0] == nt || this->can_be_nil(prod_temp[0].nonTerminal().name))
                {
                  std::unordered_set<std::string> temp = followInner(other_nt.name);
                  follows.insert(temp.begin(), temp.end());
                }

                break;
              }
              else
              {
                auto& next_symbol = prod_temp[1];
                if (next_symbol.isNonTerminal())
                {
                  const NonTerminal& next_nt = next_symbol.nonTerminal();
                  for (const std::vector<std::string>& x : this->first(next_nt.name))
                  {
                    for (const std::string& item : x)
                      follows.insert(item);
                  }

                  if (!this->can_be_nil(next_nt.name))
                    break;
                }
                else
                {
                  follows.insert(next_symbol.str());
                  break;
                }
              }

              prod_temp.erase(prod_temp.begin());
            }
          }
        }
      }
    }

    follows.erase("Nil");

    return follows;
  };

  return followInner(_name);
}

std::string dumpProduction(const Production& production)
{
  std::string retval;
  for (const auto& item : production)
  {
    if (item.isStr())
      retval += item.str();
    else
      retval += "NT(" + item.nonTerminal().name + ")";

    retval += " ";
  }

  if (!retval.empty())
    retval.resize(retval.size() - 1);

  return retval;
}